solver/llbsolver/provenance/capture.go

```diff
@@ -152,7 +152,9 @@ func (c *Capture) AddImage(i ImageSource) {
 				return
 			}
 			if v.Platform != nil && i.Platform != nil {
-				if v.Platform.Architecture == i.Platform.Architecture && v.Platform.OS == i.Platform.OS && v.Platform.Variant == i.Platform.Variant {
+				// TODO: OSFeatures, which is a string slice. Order probably does not matter, but should check the matcher.
```

Comment:
I am not sure that there is currently any Frontend that can request a particular set of OSFeatures. This particular field may be useful when trying to fetch a suitable image to run via a CRI interface, but I am not sure if this is something that will be relevant to buildkit. In most cases the `FROM` stanza will probably target the platform, and in the case of Windows, the `OSVersion` will be leveraged by the platform matcher. But `OSFeatures` seems like a layer above what buildkit needs to concern itself with. It may be something that we could update if we add extra OS features (?) via `Install-WindowsFeature` or `Enable-WindowsOptionalFeature`, but otherwise, buildkit will probably just need to copy this field from the base image.

I am not sure if there is a frontend that can update that field. Not even sure if LLB is aware of it in any meaningful way.

Reply:
I believe `Install-WindowsFeature` doesn't work in containers.

Reply:
So if this is the only blocker in terms of marking this branch as "ready for review", I think it's safe to do so. We can of course continue to discuss it in more detail as reviews come in.

I don't think this will break anything on Linux either.

Reply:
> I believe `Install-WindowsFeature` doesn't work in containers.

Have not tried it but you may be right as it required DISM. Although DISM does have the ability to enable features in an online Windows image. In any case, I don't see any reason to mutate that field outside of those two scenarios. So I think we can go ahead with this PR in it's current form.

Reply:
My main concern here is that _if_ any image uses `OSFeatures` (I've never seen one, but I've never looked) this will potentially early-out and leave an image out of the provenance.

As a temporary workaround, I could make this match if both have 0-length `OSFeatures`, and at worst, an image could be listed twice even if its platform etc. are the same. I haven't looked around to work out how you could actually reach this code, since if OSFeatures is different, wouldn't the content hash also be different? (That applies to this whole block, actually. I need to look at the context more, this code clearly does not do what I assume it does.)

----

I just checked [image-spec](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/config.md#properties) and it turns out OSFeatures (`os.features` in the data) is actually requirements on the _host_, not a description of the contents. A quick search with Google turned up nothing on the 'net about it, but poking around GitHub shows that the [use-case is when a container cannot be hosted on Nano Server](https://github.com/opencontainers/image-spec/pull/632#issuecomment-294541618) (which the spec also hints at), and LTSC 2019 onwards only supported Nano Server as a container image base anyway; [bare-metal and VM support was removed in 2017](https://techcommunity.microsoft.com/t5/core-infrastructure-and-security/nano-server-2018-update/ba-p/259512).

That last link says that server core images actually _had_ an `OSFeatures` value in 2017. I should check if that's still true, because if so, then we actually _can_ get image indexes with multiple images separated only by `OSFeatures`, i.e., a nanoserver-based image and a server-core-based image in the same index... PowerShell-Docker and DotNet-Docker are possible candidates for this. A quick check of the docs shows PowerShell-Docker only includes server core in their multi-arch image tag `latest` (not nanoserver) and DotNet apparently doesn't include _either_ in their `8.0` tag, but use nanoserver for their multi-arch image tag `latest`. The mind boggles, but that rules out the two projects I know that build multi-arch images and also support multiple Windows versions and OS variants.

I don't recall seeing any code in any platform matchers ([containerd windows-specific matcher](https://github.com/containerd/containerd/blob/332fbd907e6e9e867db913c92ff7abf07f4ce30b/platforms/defaults_windows.go#L48-L65) + [containerd generic matcher](https://github.com/containerd/containerd/blob/332fbd907e6e9e867db913c92ff7abf07f4ce30b/platforms/platforms.go#L151-L156), [moby](https://github.com/moby/moby/blob/dcf7287d647bcb515015e389df46ccf1e09855b7/distribution/pull_v2_windows.go#L136-L153)) that would support this checking though. containerd would certainly be doing the wrong thing in this case, since [its default platform matcher does not include the `win32k` feature](https://github.com/containerd/containerd/blob/332fbd907e6e9e867db913c92ff7abf07f4ce30b/platforms/defaults_windows.go#L30-L40), and so should have been rejecting anything built from a server-core-based image in 2017. (The mooted image index from the last paragraph probably selects the one that appears first. If culling version-unsuitable images is not a stable sort, that could vary from run to run. 🚨🚨🚨)

So until Platform Matchers are able to make a sensible choice using `OSFeatures` to pick an image, e.g., "when the choice is possible, exclude any with host-unsupported features and then prefer the one with the most vowels", we can probably ignore `OSFeatures` in this check. (I think this was the only one with semantics other than "copy `Platform` object... I'll check that too.)

Maybe OCI Image Spec 1.2 can deprecate the field, if it's no longer rational. And someday later we can have a new `HostFeatures` field if we need it, since `OSVersion` now describes the OS _inside_ the container, so `OSFeatures` is an annoyingly misleading name.

tl;dr TODOs:
[] rewrite the TODO to summarise the important bits of the above, and include the zero-length check for an easy win.
[] check the other changes to ensure that if they are not just "copy `Platform` object", they have similar no-op semantics and a TODO. (I think this means dropping the `*OSFEATURES` env vars exposure too.)
[] erk... copying the string slice isn't what the current copy code does, I suspect it actually takes a new reference to it. I should fix that, even though they're likely empty and no one else noticed this either. (Or I'm wrong, and the current code is fine, and I'm thinking of Python lists)
[] check if Server Core images (or any windows/\*\* base images on MCR, really) actually populate `OSFeatures`
[] depending on the above, propose deprecating `os.features` from the OCI Image Spec as unused and (now) ill-named. It's already only a `SHOULD`, so let's make it a `DON'T`.

----

tl;dr: You've gone to far. Go one paragraph earlier, this is a side-note about `Install-Windows-Feature` and `dism`.

I'm of the understanding you are supposed to be add features to a container image. I recall there being instructions for adding the OpenSSH Server feature, but it required creating an offline installer package for the feature, and if I remember correctly, last time I tried it I couldn't find the relevant ISO on my MSDN Professional or Windows Server Insiders accounts, and I don't think I have an active MSDN subscription anymore either.

Huh, apparently [I'm quoting myself there](https://github.com/microsoft/Windows-Containers/issues/119). Also, [Fonts](https://github.com/microsoft/Windows-Containers/issues/355#issuecomment-1597872938) and [Server-Media-Foundation](https://github.com/microsoft/Windows-Containers/issues/159#issuecomment-1045390706) are intended to be installable, or at least not getting "we don't support that" feedback from MS members.

Reply:
> So until Platform Matchers are able to make a sensible choice using `OSFeatures` to pick an image, e.g., "when the choice is possible, exclude any with host-unsupported features and then prefer the one with the most vowels", we can probably ignore `OSFeatures` in this check. (I think this was the only one with semantics other than "copy `Platform` object... I'll check that too.)

This is my preference as well. No use blocking this due to `OSFeatures`. Right now, nothing works. With this patch, most of it will work. I'd rather we do an unplug-the-server-and-see-who-yells test than not continue at all. A TODO/NOTE here should be enough.

>  ☐ erk... copying the string slice isn't what the current copy code does, I suspect it actually takes a new reference to it. I should fix that, even though they're likely empty and no one else noticed this either. (Or I'm wrong, and the current code is fine, and I'm thinking of Python lists)

indeed, that is python :smile:. Lists in go are [not passed by reference](https://go.dev/play/p/u1oS0nAAZfx), unless you pass an actual pointer to the list. I find myself writing python in go projects as well when I context switch a lot. 

> Edit: OMG, [the font installation instructions](https://techcommunity.microsoft.com/t5/itops-talk-blog/adding-optional-font-packages-to-windows-containers/ba-p/3559761) have a horrible creative workaround for lack of RUN --mount. I don't want to spoil the surprise, but NET USE is involved. (Actually, I'm not sure you could achieve what they are doing with RUN --mount, unless you somehow got C:\Windows\WinSxS to be inside the build context.)

I will pretend I didn't see that. Also, will look into the `--mount` code to see if we can add that to Windows and hopefully spare someone the mental anguish of resorting to a file share to emulate mounts.

Reply:
> > ☐ erk... copying the string slice isn't what the current copy code does, I suspect it actually takes a new reference to it. I should fix that, even though they're likely empty and no one else noticed this either. (Or I'm wrong, and the current code is fine, and I'm thinking of Python lists)
> 
> indeed, that is python 😄. Lists in go are [not passed by reference](https://go.dev/play/p/u1oS0nAAZfx), unless you pass an actual pointer to the list. I find myself writing python in go projects as well when I context switch a lot.

https://go.dev/play/p/MkOIPlk-K6Z is a behaviour I'm concerned about, and it looks like they _do_ simple-assign by reference. And it looks like even creating a new slice with `[:]`, either I was able to mutate the string itself, or it optimised that back into normal assignment (which copies the internal slice pointer/length); that would have worked in Python.

In your Go example, appending to `sl` deduplicates it from `aa` because it needs to be grown, see https://go.dev/play/p/8a9plAYrn4i where `sl` is a reference before it's appended to, but not afterwards.